pages/intelligence.md

```diff
@@ -0,0 +1,137 @@
+on-disk representation of the index.
+
+Such a representation will need to take the following into account:
+
+  * Branches can while the language server is running, invalidating large portions of the cache
+  * Project level search and replace can affect large numbers of files at once
+  * The versions of elixir and erlang can change via `asdf` whenever the user wants
+
+Basicallly, the project's assets can change at any time, and it's not
+guaranteed that the language server will be able to detect all of the
+changes, so the language server needs to treat this cache with some
+skepticism and constantly "check its work".
+
+## Cache layout
+A cache entry will have the following format:
+
+```elixir
+
+@type source :: Lexical.uri()
+@entry_type :: :otp | :elixir | :dependency | :project
+
+@type entry %{
+  type: entry_type(),
+  source: source(),
+  entry_version: String.t(),
+  added_at: DateTime.t(),
+  value: term()
+}
+```
+
+### Invalidation
+When lexical starts the first time, it will need to index the
+following entities in the following order:
+
+  1. The elixir release
+  2. All project dependencies
+  3. The project's source code and built artifacts
+
+The index needs to take into account which stage added it, and include an appropriate
+version when adding an entry. The versions are as follows:
+
+| Entry Type  | Entry Version |
+| ----------- | --------------|
+| :otp        | [This insanity](https://stackoverflow.com/questions/9560815/how-to-get-erlangs-release-version-number-from-a-shell/34326368#34326368)|
+| :elixir     | `System.version()`|
+| :dependency | `Application.spec(dep_atom)[:vsn]`|
+| :project    | `md5(file contents)` |
+
+Entries in the cache should be replaced and not updated, and we should
+have a mechanism to sweep old entries from the cache (for example, if
+a dependecy version is old and hasn't been updated in two weeks, we
+should remove it). This invalidation is going to be one of the tricky
+parts of the cache, and we should be aggressive with it, as
+re-indexing things is likely to be less bad than having an
+ever-growing cache on disk.  We need to think more about this and come
+up with something that strikes a balance between efficiency and
+on-disk storage.
+
+# Indexing
+Indexing and searching go hand-in-hand, you need to make an index in a
+format that's useful, and as such, we'll approach this at a later
+date. In the interim, we can probably get away with a linear search;
+there aren't _that many_ things that we're indexing and elixir is kind
+of fast for this kind of thing.
+
+# Recognition
+When we do things like "go to definition" or "find references" we need
+to understand the context beneath the cursor to determine _what_ to
+look for. However, this can be somewhat challenging, because elixir
+has ambiguity in its syntax. For example:
+
+```elixir
+foo
+```
+
+can be a variable or a function call. Our recognition engine needs to
```

Comment:
> can be a variable or a function call

thanks Ruby syntax? :p 

Funny given that now parenthesis are more or less enforced by `mix format` for function calls

Reply:
well, not for DSLs, which we need to support. Things get really hairy there.

Reply:
True. I would be fine with kicking that problem down the road until everything else works great, and only at that point start thinking about abstractions that work for the DSL problem.